returnn/tf/layers/basic.py

```diff
@@ -854,44 +858,86 @@ class SliceNdLayer(_ConcatInputLayer):
     self.start = start
-    assert start.output.have_batch_axis() and start.output.batch_shape == (None,)
-    start = start.output.get_placeholder_as_batch_major()
+    start_data = start.output.copy_as_batch_major()  # e.g. (B,) or (B,T)
+    start_t = start_data.placeholder
     if size is None:
       if seq_lens is None:
-        size = tf.maximum(tf.reduce_max(x.batch_shape[1] - start), 0)
+        size = tf.maximum(tf.reduce_max(x.batch_shape[1] - start_t), 0)  # scalar
       else:
-        size = tf.maximum(tf.reduce_max(seq_lens - start), 0)
+        # make seq_lens compatible with start_t
+        seq_lens = expand_multiple_dims(  # e.g. (B,) or (B,1)
+          x=seq_lens,
+          axes=[-1] * (len(start_t.shape) - len(seq_lens.shape)))
+        size = tf.maximum(tf.reduce_max(seq_lens - start_t), 0)  # scalar
       if min_size is not None:
         size = tf.maximum(size, min_size)
-    self.size = size
-    start = tf.expand_dims(start, axis=1)  # (B, T)
-    slices = slice_nd(x.placeholder, start=tf.cast(start, tf.int32), size=size)  # (B,size, ...)
-    if seq_lens is not None:
-      mask = tf.greater_equal(tf.range(size)[None, :] + start, seq_lens[:, None])  # (B,T)
-      mask = expand_multiple_dims(mask, list(range(2, x.batch_ndim)))
-      slices = where_bc(mask, tf.zeros_like(slices), slices)
-    size_placeholder = x.size_placeholder.copy()
-    if isinstance(size, tf.Tensor):
-      size_placeholder[0] = tf.maximum(seq_lens - tf.reshape(start, tf.shape(seq_lens)), 0)
+    # build Data object for the position argument of GatherLayer
+    indices_data = start_data.copy_template(name="%s_gather_indices" % self.name)
+    if isinstance(size, int):
       tag = DimensionTag(
+        kind=DimensionTag.Types.Spatial,
         description="sliced-time:%s" % self.get_absolute_name(),
-        kind=DimensionTag.Types.Spatial, batch=self.output.batch)
-      tag.set_tag_on_size_tensor(size_placeholder[0])
+        batch=self.output.batch,
+        dimension=size)
     else:
-      assert isinstance(size, int)
-      size_placeholder.pop(0, None)  # static time axis
-    self.output.size_placeholder = size_placeholder
-    self.output.placeholder = slices
+      # in this case, size is not known before runtime and becomes dynamic
+      if len(seq_lens.shape) == 1:
+        dyn_size = tf.maximum(seq_lens - start_t, 0)  # (B,)
+      else:
+        # in this case, we would normally get a dynamic size of shape (B,T) for the slices
+        # in order to get shape (B,) instead, we reduce all other axes except the batch axis
+        reduce_axes = range(1, len(seq_lens.shape))
+        dyn_size = tf.maximum(tf.reduce_max(seq_lens - start_t, axis=reduce_axes), 0)  # (B,)
+      tag = DimensionTag(
+        kind=DimensionTag.Types.Spatial,
+        description="sliced-time:%s" % self.get_absolute_name(),
+        batch=self.output.batch,
+        dyn_size=dyn_size)
```

Comment:
Same here. Take the existing dim tag.
And then assign the `dyn_size` here:
`tag.dyn_size = dyn_size`.